File: Experiment 1/MemoryBenchmark.java
public class MemoryBenchmark {
    public static void main(String[] args){
        Integer size = Integer.parseInt(args[0]);
        Integer loops = 1000;
        long[] creationTimes = new long[loops];
        long executionTimeStart = System.currentTimeMillis();

        for (int i = 0; i < loops; i++){
            creationTimes[i] = System.nanoTime();
            int[] array = new int[size];
            for (int j = 0; j < size; j++){
                array[j] = j;
            }
            creationTimes[i] = System.nanoTime() - creationTimes[i];
        }

        long executionTime = System.currentTimeMillis() - executionTimeStart;

        // sort the creation times to calculate the mean
        for (int i = 0; i < loops - 1; i++){
            for (int j = 0; j < loops - i - 1; j++){
                if (creationTimes[j] > creationTimes[j + 1]){
                    long temp = creationTimes[j];
                    creationTimes[j] = creationTimes[j + 1];
                    creationTimes[j + 1] = temp;
                }
            }
        }
        // get the mean creation time 
        long meanCT = creationTimes[loops / 2];
        
        System.out.println("Median loop time " + meanCT + "ns");
        System.out.println("Execution time " + executionTime + "ms");
    }
}

File: Experiment 3/Experiment 3.0/MemoryBenchmark.cpp
#include <chrono>
#include <string>
#include <iostream>
#include <algorithm>

int main(int argc, char *argv[]) {
    int size = std::stoi(argv[1]);
    int loops = 1000;
    // warm up the program
    for (int i = 0; i < loops; i++) {
        int* array = new int[size];
        for (int j = 0; j < size; j++) {
            array[j] = j;
        }
        delete[] array;
    }

    std::chrono::duration<long int, std::ratio<1, 1000000000>>* creationTimes = new std::chrono::duration<long int, std::ratio<1, 1000000000>>[loops];
    auto executionTimeStart = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < loops; i++) {
        auto creationTime = std::chrono::high_resolution_clock::now();
        int* array = new int[size];
        for (int j = 0; j < size; j++) {
            array[j] = j;
        }
        creationTimes[i] = std::chrono::high_resolution_clock::now() - creationTime;
        delete[] array;
    }

    auto executionTime = std::chrono::high_resolution_clock::now() - executionTimeStart;
    std::sort(creationTimes, creationTimes + loops);
    std::chrono::duration<long int, std::ratio<1, 1000000000>> meanCT = creationTimes[loops / 2];

    std::cout << "Median loop time " << meanCT.count() << "ns" << std::endl;
    std::cout << "Execution time " << std::chrono::duration_cast<std::chrono::milliseconds>(executionTime).count() << "ms" << std::endl;
    return 0;
}
